core/src/main/java/fr/sncf/osrd/infra/implementation/reservation/ReservationInfraBuilder.java

```diff
@@ -204,48 +207,59 @@ private void validateRoute(ReservationRouteImpl route, RJSRoute rjsRoute) {
-            if (g.adjacentEdges(prev.track.getEdge()).contains(next.track.getEdge()))
-                continue;
-            var prevNode = GraphHelpers.nodeFromEdgeEndpoint(
-                    g,
-                    prev.track.getEdge(),
-                    Direction.endEndpoint(prev.track.getDirection())
-            );
-            var nextNode = GraphHelpers.nodeFromEdgeEndpoint(
-                    g,
-                    next.track.getEdge(),
-                    Direction.startEndpoint(next.track.getDirection())
-            );
-            var connecting = g.edgeConnecting(prevNode, nextNode);
-            if (connecting.isEmpty())
-                connecting = g.edgeConnecting(nextNode, prevNode);
-            if (connecting.isEmpty())
-                throw new DiscontinuousRoute(rjsRoute.id, prev.track.getEdge().getID(), next.track.getEdge().getID());
-            var branchEdge = connecting.get();
-            var dir = g.incidentNodes(branchEdge).nodeU().equals(prevNode) ? FORWARD : BACKWARD;
-            res.add(i, new TrackRangeView(0, 0, diTrackInfra.getEdge(branchEdge, dir)));
+        var direction = Direction.fromEdgeDir(rjsRoute.entryPointDirection);
+        var entryDetector = diTrackInfra.getDetectorMap().get(rjsRoute.entryPoint.id.id);
+        assert entryDetector != null;
+        var exitDetector = diTrackInfra.getDetectorMap().get(rjsRoute.exitPoint.id.id);
+        assert exitDetector != null;
+        var edge = diTrackInfra.getEdge(entryDetector.getTrackSection().getID(), direction);
+        var position = entryDetector.getOffset();
+        var exitEdge = diTrackInfra.getEdge(exitDetector.getTrackSection().getID(), FORWARD).getEdge();
+        var g = diTrackInfra.getDiTrackGraph();
+
+        while (edge.getEdge() != exitEdge) {
+            // Add track range
+            double endPosition = edge.getDirection() == FORWARD ? edge.getEdge().getLength() : 0.;
+            res.add(new TrackRangeView(position, endPosition, edge));
+
+            // Find next track
+            var endpointPair = g.incidentNodes(edge);
+            var neighbors = g.outEdges(endpointPair.target());
+
+            if (neighbors.isEmpty())
+                throw new DiscontinuousRoute(rjsRoute.id);
+
+            var firstNeighbor = neighbors.iterator().next();
+            if (!(firstNeighbor.getEdge() instanceof SwitchBranch)) {
+                assert neighbors.size() == 1;
+                // Simple track link
+                edge = firstNeighbor;
+            } else {
+                // Switch link
+                var aSwitch = ((SwitchBranch) firstNeighbor.getEdge()).getSwitch();
+                var switchGroup = rjsRoute.switchesDirections.get(aSwitch.getID());
+                var switchBranches = aSwitch.getGroups().get(switchGroup);
+                var newEdge = edge;
+                for (var neighbor : neighbors) {
+                    if (switchBranches.contains(neighbor.getEdge())) {
```

Comment:
Intellij linter isn't happy about this, we look for a `TrackEdge` in a collection of `SwitchBranches`. We could ignore it, but we can also assert that we are indeed dealing with `SwitchBranches`